### ISSUE TITLE
Allow models.Span.location to accept categorical values

### ISSUE THREAD

**Issue by jorgecarleitao** [bokeh/bokeh#8592]
This suggestion is based on [this SO question](https://stackoverflow.com/questions/54415460/how-to-plot-a-vertical-line-on-a-bar-plot-in-bokeh). 

The proposal here is to allow something like `vline = Span(location='Apples', dimension='height')` when the `x_range` is a categorical axis.

Based on the first example of the user-guide:

```
from bokeh.io import show, output_file
from bokeh.plotting import figure
from bokeh.models import Span

output_file("bars.html")

fruits = ['Apples', 'Pears', 'Nectarines', 'Plums', 'Grapes', 'Strawberries']
counts = [5, 3, 4, 2, 4, 6]

p = figure(x_range=fruits, plot_height=250, title="Fruit Counts",
           toolbar_location=None, tools="")

p.vbar(x=fruits, top=counts, width=0.9)

# SEE THIS LINE
p.renderers.extend([Span(location='Apples', dimension='height')])

p.xgrid.grid_line_color = None
p.y_range.start = 0

show(p)
```

I am trying to add a vertical line to a bar plot whose x-range are categories. However, this does not seem to be possible, as this raises an error "ValueError: expected a value of type Real, got Apples of type str". `location='Apples'` does not work as intended as it expected a number.

One solution is to use

```
index = p.x_range.factors.index("Apples")
delta = (p.x_range.end - p.x_range.start)/p.x_range.factors.length;
location = delta/2 + index;
```

but this IMO is a workaround. Would be nice for Bokeh to support this out of the box.

**Comment by bryevdv** [comment]
The proximate issue here is that `Span` has `location = Float(...)` which rejects anything that is not s number. Glyphs handle categorical factors in CDS columns because they have e.g `x = NumberSpec(...)`. (As an aside I think that means that glyphs can *only* handle factors from CDS columns and not as literal single values). In order to support factors for `Span` the property type of `location` will need to be widened. A quick experiment might try 
```
location = Either(String, Float)
```
to see if the BokehJS side can already happen to deal with factors. But to allow nested categories, a more precise property will be needed.

**Comment by mattpap** [comment]
Another similar issue: https://github.com/bokeh/bokeh/issues/8482. We need to make all annotations aware of categoricals.

**Comment by gshpychka** [comment]
Is there any progress on this?

**Comment by bryevdv** [comment]
@gshpychka There is never any need to ask this sort of question. If there were any work or updates, they would be recorded here on the issue.

**Comment by gshpychka** [comment]
@bryevdv thanks, I figured there was some other place where you decide what to prioritize, what to include in which release, etc, so I was hoping to hear if there was any kind of decision made w.r.t a timeline for this.

**Comment by bryevdv** [comment]
If it is slated for an upcoming release it will get a version Milestone set

**Comment by mattpap** [comment]
This was fixed in PR #12541.
